tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -15288,6 +15342,42 @@ void DiagnoseNodeEntry(Sema &S, FunctionDecl *FD, llvm::StringRef StageName,
       S.Diags.Report(NodeDG->getLocation(), diag::note_defined_here)
           << NodeDG->getSpelling();
     }
+    // Node with NodeMaxDispatchGrid must have SV_DispatchGrid semantic.
+    if (NodeMDG) {
+      bool Found = false;
+      for (FunctionDecl::param_iterator I = FD->param_begin(),
+                                        E = FD->param_end();
+           I != E; ++I) {
+        QualType ParamType = (*I)->getType().getCanonicalType();
+
+        // Find parameter that is the node input record
+        if (hlsl::IsHLSLNodeInputType(ParamType)) {
+          // Node input records are template types
+          if (const RecordType *NodeInputRT = dyn_cast<RecordType>(ParamType)) {
+            if (const ClassTemplateSpecializationDecl *templateDecl =
+                    dyn_cast<ClassTemplateSpecializationDecl>(
+                        NodeInputRT->getDecl())) {
+
+              // Get the input record struct
+              auto &TemplateArgs = templateDecl->getTemplateArgs();
+              auto &StructArg = TemplateArgs.get(0);
+              QualType StructArgType = StructArg.getAsType();
+              CXXRecordDecl *NodeInputStructDecl = dyn_cast<CXXRecordDecl>(
+                  StructArgType->getAsStructureType()->getDecl());
+
+              // Make sure there is exactly one SV_DispatchGrid semantics
+              DiagnoseMustHaveOneDispatchGridSemantics(S, NodeInputStructDecl,
```

Comment:
How about putting the previous statement in an if condition, and the recursive call inside the if body?